e2e/test/iothub/service/DirectMethodE2ETests.cs

```diff
@@ -0,0 +1,127 @@
+            errorContext.And.ErrorCode.Should().Be(IotHubServiceErrorCode.ModuleNotFound);
+        }
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_DeviceOnline_DeviceNotSubscribedToDirectMethods()
+        {
+            // arrange
+            // device is online but device client is not explicitly opened
+            IotHubServiceClient serviceClient = TestDevice.ServiceClient;
+            await using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(_devicePrefix);
+            
+            //IotHubDeviceClient deviceClient = testDevice.CreateDeviceClient();
+            var methodInvocation = new DirectMethodServiceRequest("someDirectMethod");
+
+            // act
+            Func<Task> act = async () => await serviceClient.DirectMethods.InvokeAsync(testDevice.Id, methodInvocation);
+
+            // assert
+            ExceptionAssertions<IotHubServiceException> errorContext = await act.Should().ThrowAsync<IotHubServiceException>();
+            errorContext.And.StatusCode.Should().Be(System.Net.HttpStatusCode.NotFound);
+            errorContext.And.ErrorCode.Should().Be(IotHubServiceErrorCode.DeviceNotOnline);
+        }
+
+        [TestMethod]
+        [Timeout(TestTimeoutMilliseconds)]
+        public async Task DirectMethodsClient_DeviceOnline_ExceedsRequestTimeout()
+        {
+            // arrange
+            IotHubServiceClient serviceClient = TestDevice.ServiceClient;
+            await using TestDevice testDevice = await TestDevice.GetTestDeviceAsync(_devicePrefix);
+            const int timeoutInSeconds = 3;
+
+            var methodInvocation = new DirectMethodServiceRequest("someDirectMethod")
+            {
+                ConnectionTimeout = TimeSpan.FromSeconds(timeoutInSeconds),
+                ResponseTimeout = TimeSpan.FromSeconds(timeoutInSeconds),
+            };
+            IotHubDeviceClient deviceClient = testDevice.CreateDeviceClient();
+            await testDevice.OpenWithRetryAsync().ConfigureAwait(false);
+
+            // act
+            await deviceClient
+            .SetDirectMethodCallbackAsync(
+                async (methodRequest) =>
+                {
+                    // force a timeout
+                    await Task.Delay(timeoutInSeconds * 1000 * 2).ConfigureAwait(false);
+                    var response = new DirectMethodResponse(200);
+                    return response;
+                })
+            .ConfigureAwait(false);
+
+            // assert
+            Func<Task> act = async() => await serviceClient.DirectMethods.InvokeAsync(testDevice.Id, methodInvocation);
+            ExceptionAssertions<IotHubServiceException> response = await act.Should().ThrowAsync<IotHubServiceException>();
+            response.And.StatusCode.Should().Be(System.Net.HttpStatusCode.BadRequest);
+            response.And.ErrorCode.Should().Be(IotHubServiceErrorCode.ArgumentInvalid);
```

Comment:
This is surprising. Do we get the same error over AMQP?

Reply:
It looks like it. I can ping a service team member to see if they have insight, but it might not be something we can change.